### PR TITLE
Fix dissect ingest processor parsing empty brackets failed

### PR DESCRIPTION
**Pull request by gaobinlong**
### Description
Fix the bug of dissect ingest processor parse empty brackets failed, and add some unit test code and yaml rest test.
The bug is caused by that when walking through the original string, the cursor `i` cannot move to next when one or more matches are found because i already jumps to then end of the match(i += lookAheadMatches), so `for` loop should be changed to `while` loop.  

### Related Issues
Relates to https://github.com/opensearch-project/OpenSearch/issues/7223.

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
This looks like a real bug, not sure why we didn't look at it. Maybe @msfroh?

**Comment by dblock** [comment]
@gaobinlong Apologies for letting this hang here. I've resolved some CHANGELOG conflicts, but maybe you want to rebase the change?

File: libs/dissect/src/main/java/org/opensearch/dissect/DissectParser.java [hunk 1]
```diff
@@ -284,7 +284,11 @@ public Map<String, String> parse(String inputString) {
                         delimiter = dissectPair.getDelimiter().getBytes(StandardCharsets.UTF_8);
                         // i is always one byte after the last found delimiter, aka the start of the next value
                         valueStart = i;
+                    } else {
```
Comment (dblock) [inline libs/dissect/src/main/java/org/opensearch/dissect/DissectParser.java:287]: Any chance `i` doesn't change above? I would at least assert or something, after all we *are* fixing an infinite loop here.

Comment (gaobinlong) [reply libs/dissect/src/main/java/org/opensearch/dissect/DissectParser.java:290]: Seems not, I've added an assertion for this.

**Comment by gaobinlong** [comment]
Both the failed test [org.opensearch.persistent.PersistentTasksExecutorFullRestartIT.testFullClusterRestart](https://github.com/opensearch-project/OpenSearch/issues/5145) and [org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testUpdateOperations ](https://github.com/opensearch-project/OpenSearch/issues/8279) seem to be flaky tests.

**Comment by dblock** [comment]
> Both the failed test [org.opensearch.persistent.PersistentTasksExecutorFullRestartIT.testFullClusterRestart](https://github.com/opensearch-project/OpenSearch/issues/5145) and [org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testUpdateOperations ](https://github.com/opensearch-project/OpenSearch/issues/8279) seem to be flaky tests.

Find the GitHub issue for those/open one if it doesn't exist. Force push/rebase to retry gradle checks until a green one.

**Comment by gaobinlong** [comment]
The flaky tests have related issues: https://github.com/opensearch-project/OpenSearch/issues/5145, https://github.com/opensearch-project/OpenSearch/issues/10026. Now all checks have passed yet.

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
There was a CHANGELOG conflict, so I resolved it changing the language slightly to "Fix failure in dissect ingest processor parsing empty brackets", lmk if that's incorrect.

**Pull request merged**